python/dask_cudf/dask_cudf/sorting.py

```diff
@@ -0,0 +1,381 @@
+    if isinstance(df, (Series, Index)):
+        df = df.to_frame()
+    assert isinstance(df, DataFrame)
+    final_type = df._meta._constructor
+
+    # Create metadata
+    meta = df._meta_nonempty.quantile(q=q)
+
+    # Define final action (create df with quantiles as index)
+    def finalize_tsk(tsk):
+        return (final_type, tsk, q)
+
+    return_type = DataFrame
+
+    # pandas/cudf uses quantile in [0, 1]
+    # numpy / cupy uses [0, 100]
+    qs = np.asarray(q)
+    token = tokenize(df, qs)
+
+    if len(qs) == 0:
+        name = "quantiles-" + token
+        empty_index = gd.Index([], dtype=float)
+        return Series(
+            {
+                (name, 0): final_type(
+                    {col: [] for col in df.columns},
+                    name=df.name,
+                    index=empty_index,
+                )
+            },
+            name,
+            df._meta,
+            [None, None],
+        )
+    else:
+        new_divisions = [np.min(q), np.max(q)]
+
+    name = "quantiles-1-" + token
+    val_dsk = {
+        (name, i): (_quantile, key, qs)
+        for i, key in enumerate(df.__dask_keys__())
+    }
+
+    name2 = "quantiles-2-" + token
+    merge_dsk = {
+        (name2, 0): finalize_tsk(
+            (merge_quantiles, qs, [qs] * df.npartitions, sorted(val_dsk))
+        )
+    }
+    dsk = toolz.merge(val_dsk, merge_dsk)
+    graph = HighLevelGraph.from_collections(name2, dsk, dependencies=[df])
+    return return_type(graph, name2, meta, new_divisions)
+
+
+def quantile_divisions(df, by, npartitions):
+    qn = np.linspace(0.0, 1.0, npartitions + 1).tolist()
+    divisions = _approximate_quantile(df[by], qn).compute()
+    columns = divisions.columns
+
+    # TODO: Make sure divisions are correct for all dtypes..
+    if (
+        len(columns) == 1
+        and df[columns[0]].dtype != "object"
+        and not is_categorical_dtype(df[columns[0]].dtype)
+    ):
+        dtype = df[columns[0]].dtype
+        divisions = divisions[columns[0]].astype("int64")
+        divisions.iloc[-1] += 1
+        divisions = sorted(
+            divisions.drop_duplicates().astype(dtype).values.tolist()
+        )
+    else:
+        for col in columns:
+            dtype = df[col].dtype
+            if dtype != "object":
+                divisions[col] = divisions[col].astype("int64")
+                divisions[col].iloc[-1] += 1
+                divisions[col] = divisions[col].astype(dtype)
+            else:
+                divisions[col].iloc[-1] = chr(
+                    ord(divisions[col].iloc[-1][0]) + 1
+                )
+        divisions = divisions.drop_duplicates()
+    return divisions
+
+
+def sort_values(
+    df,
+    by,
+    max_branch=None,
+    divisions=None,
+    set_divisions=False,
+    ignore_index=False,
+):
+    """ Sort by the given list/tuple of column names.
+    """
+    npartitions = df.npartitions
+    if isinstance(by, str):
+        by = [by]
```

Comment:
Does Dask enforce column names are strings? Otherwise this can be an integer or other types as well.

Reply:
You're correct - The str assumption is a bad idea here.